src/lib/Linearize.hs

```diff
@@ -143,10 +143,8 @@ liftTangentM args m = liftSubstReaderT $ lift11 $ runReaderT1 args m
 
 isTrivialForAD :: Expr o -> PrimalM i o Bool
 isTrivialForAD expr = do
-  trivialTy  <- (maybeTangentType <$> getType expr) >>= \case
-    Nothing -> return False
-    Just tTy -> isSingletonType tTy
-  hasActiveEffs <- exprEffects expr >>= \case
+  trivialTy  <- any isSingletonType . maybeTangentType <$> getType expr
```

Comment:
FWIW I find this use of `any` a bit confusing. The previous explicit statement was clearer, because one didn't have to think what is the neutral element of the "any" monoid.

Reply:
Clarified.